### PR TITLE
Add Notification objects to the notification system.

### PR DESCRIPTION
**Pull request by miterosan**

### CODE REVIEW

**Review by peppy — CHANGES_REQUESTED** [review]
needs further explanation

File: osu.Game/Screens/Play/HUDOverlay.cs [hunk 1]
```diff
@@ -72,10 +73,9 @@ private void load(OsuConfigManager config, NotificationOverlay notificationOverl
             {
                 hasShownNotificationOnce = true;
 
-                notificationOverlay?.Post(new SimpleNotification
-                {
-                    Text = @"The score overlay is currently disabled. You can toggle this by pressing Shift+Tab."
-                });
+                notificationOverlay?.Post(new SimpleNotificationContainer(
```
Comment (peppy) [inline osu.Game/Screens/Play/HUDOverlay.cs]: at very least, this should not be a thing.

NotificationOverlay should know about the mapping between models and drawables, so we should just be posting the `new Notification()` and there should be type checks (switch in c#7) to create the drawable versions.

**Comment by miterosan** [comment]
https://i.imgur.com/wlbszUy.png

File: osu.Game.Tests/Visual/TestCaseNotificationOverlay.cs [hunk 2]
```diff
@@ -91,19 +96,58 @@ private void sendProgress2()
...
+            };
+            restartNotification.OnActivate += () => manager.Post(completeNotification);
+
+            bool downloadCompleted = false;
+            var downloadNotification = new ProgressNotification(@"Downloading update...")
+            {
+                NotificationIcon = new NotificationIcon(
+                    FontAwesome.fa_upload,
+                    ColourInfo.GradientVertical(osuColour.YellowDark, osuColour.Yellow)
+                )
```
Comment (smoogipoo) [inline osu.Game.Tests/Visual/TestCaseNotificationOverlay.cs]: Inline the construction of this icon.

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 3]
```diff
@@ -236,18 +234,22 @@ public void Download(BeatmapSetInfo beatmapSetInfo, bool noVideo = false)
...
-                {
-                    Icon = FontAwesome.fa_superpowers,
-                    Text = "You gotta be a supporter to download for now 'yo"
-                });
+                PostNotification?.Invoke(
+                    new Notification(
+                        "You gotta be a supporter to download for now 'yo",
+                        FontAwesome.fa_superpowers
+                    )
+                );
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/BeatmapManager.cs]: Inline this

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 4]
```diff
@@ -236,18 +234,22 @@ public void Download(BeatmapSetInfo beatmapSetInfo, bool noVideo = false)
...
                 return;
             }
 
-            ProgressNotification downloadNotification = new ProgressNotification
-            {
-                Text = $"Downloading {beatmapSetInfo.Metadata.Artist} - {beatmapSetInfo.Metadata.Title}",
-            };
+            ProgressNotification downloadNotification = new ProgressNotification(
+                $"Downloading {beatmapSetInfo.Metadata.Artist} - {beatmapSetInfo.Metadata.Title}"
+            );
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/BeatmapManager.cs]: Inline

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 5]
```diff
@@ -236,18 +234,22 @@ public void Download(BeatmapSetInfo beatmapSetInfo, bool noVideo = false)
...
+
+            downloadNotification.FollowUpNotifications.Add(
+                new Notification($"Downloading {beatmapSetInfo.Metadata.Artist} - {beatmapSetInfo.Metadata.Title} finished")
+            );
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/BeatmapManager.cs]: Inline, basically just go through this PR and inline all of these. Makes it easier to follow imo.

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 6]
```diff
@@ -647,7 +648,7 @@ public void DeleteAll()
 
             if (maps.Count == 0) return;
 
-            var notification = new ProgressNotification
+            var notification = new ProgressNotification("")
```
Comment (smoogipoo) [inline osu.Game/Beatmaps/BeatmapManager.cs]: How about making the constructor parameter optional? Would fit well with the constructor of `Notification` too.

File: osu.Game/Notifications/Notification.cs [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        public Bindable<NotificationIcon> NotificationIconBinding { get; } = new Bindable<NotificationIcon>(new NotificationIcon());
+        public Bindable<ColourInfo> BackgroundColourBinding { get; } = new Bindable<ColourInfo>(OsuColour.FromHex("FFF"));
+        public Bindable<string> TextBinding { get; } = new Bindable<string>();
+
+        /// <summary>
+        /// An event, that gets triggered when the user activates the notification.
+        /// </summary>
+        public event Action OnActivate;
```
Comment (smoogipoo) [inline osu.Game/Notifications/Notification.cs]: How about renaming this to `Action`? Same as `ClickableContainer`.

File: osu.Game/Notifications/Notification.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+        /// <summary>
+        /// Activates this notification.
+        /// </summary>
+        public void Activate()
```
Comment (smoogipoo) [inline osu.Game/Notifications/Notification.cs]: Along with the above, perhaps rename this to `TriggerAction`.

File: osu.Game/Notifications/NotificationIcon.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+        /// <summary>
+        /// The icon displayed in the notification.
+        /// </summary>
+        public FontAwesome Icon { get; }
+
+        /// <summary>
+        /// Changes to background color of the icon. Defaults to grey
+        /// </summary>
+        public ColourInfo BackgroundColour { get; }
```
Comment (smoogipoo) [inline osu.Game/Notifications/NotificationIcon.cs]: These properties don't have bindings like all the others inside `Notification`.

In general, I'm not too certain about the bindings, versus just having them as fields and the notifications updating inside their `Update()`.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
I like the way you've gone with this.

File: osu.Game.Tests/Visual/TestCaseNotificationOverlay.cs (hunk 2 above)
Comment (miterosan) [reply osu.Game.Tests/Visual/TestCaseNotificationOverlay.cs]: https://i.imgur.com/eR6X2kh.png this feels to long for a line

File: osu.Game/Notifications/NotificationIcon.cs (hunk 9 above)
Comment (peppy) [reply osu.Game/Notifications/NotificationIcon.cs]: this is intentional. if you want to make changes you'd replace the `NotificationIcon` itself.

File: osu.Game/Notifications/Notification.cs (hunk 7 above)
Comment (miterosan) [reply osu.Game/Notifications/Notification.cs]: Thought of this... but the implementation required is meh.....
https://i.imgur.com/UUXoSBj.png

File: osu.Game/Beatmaps/BeatmapManager.cs (hunk 4 above)
Comment (miterosan) [reply osu.Game/Beatmaps/BeatmapManager.cs]: Same with above, line gets too long.

File: osu.Game/Overlays/Notifications/ProgressNotificationDrawable.cs [hunk 10]
```diff
@@ -145,28 +121,23 @@ private void load(OsuColour colours)
...
         {
-            switch (State)
+            switch (ProgressNotification.State)
             {
                 case ProgressNotificationState.Cancelled:
                     base.Close();
                     break;
                 case ProgressNotificationState.Active:
                 case ProgressNotificationState.Queued:
-                    if (CancelRequested?.Invoke() != false)
```
[the remark this answers is not in the document]
Comment (miterosan) [reply osu.Game/Overlays/Notifications/ProgressNotificationDrawable.cs]: https://i.imgur.com/TDcsxIG.png ?

File: osu.Desktop/Overlays/VersionManager.cs [hunk 11]
```diff
@@ -154,31 +144,45 @@ protected override void Dispose(bool isDisposing)
                     //no updates available. bail and retry later.
                     return;
 
-                if (notification == null)
+                if (notificationContainer == null)
                 {
-                    notification = new UpdateProgressNotification { State = ProgressNotificationState.Active };
```
Comment (peppy) [inline osu.Desktop/Overlays/VersionManager.cs]: honestly this feels worse than before. now you're having to pollute the local class with design-level knowledge of the drawable notification. and this isn't just the case here. every time a notification wants to be coloured this is going to be the case.

File: osu.Desktop/Overlays/VersionManager.cs [hunk 12]
```diff
@@ -140,7 +130,7 @@ protected override void Dispose(bool isDisposing)
             updateManager?.Dispose();
         }
 
-        private async void checkForUpdateAsync(bool useDeltaPatching = true, UpdateProgressNotification notification = null)
+        private async void checkForUpdateAsync(bool useDeltaPatching = true, ProgressNotification notificationContainer = null)
```
Comment (peppy) [inline osu.Desktop/Overlays/VersionManager.cs]: it's not a container though?

File: osu.Desktop/Overlays/VersionManager.cs [hunk 13]
```diff
@@ -222,44 +226,5 @@ protected override void PopOut()
         {
         }
 
-        private class UpdateProgressNotification : ProgressNotification
```
Comment (peppy) [inline osu.Desktop/Overlays/VersionManager.cs]: this was much more contained and nicer to look at.

File: osu.Game/Beatmaps/BeatmapManager.cs [hunk 14]
```diff
@@ -125,16 +125,14 @@ public BeatmapManager(Storage storage, Func<OsuDbContext> context, RulesetStore
 
         /// <summary>
         /// Import one or more <see cref="BeatmapSetInfo"/> from filesystem <paramref name="paths"/>.
-        /// This will post a notification tracking import progress.
+        /// This will post a drawableNotification tracking import progress.
```
Comment (peppy) [inline osu.Game/Beatmaps/BeatmapManager.cs]: ?

File: osu.Game/Overlays/NotificationOverlay.cs [hunk 15]
```diff
@@ -79,23 +80,36 @@ private void notificationClosed()
...
-                notification.Depth = notification.DisplayOnTop ? runningDepth : -runningDepth;
+                drawableNotification.Depth = drawableNotification.DisplayOnTop ? runningDepth : -runningDepth;
 
-                notification.Closed += notificationClosed;
+                drawableNotification.Closed += notificationClosed;
 
-                var hasCompletionTarget = notification as IHasCompletionTarget;
-                if (hasCompletionTarget != null)
-                    hasCompletionTarget.CompletionTarget = Post;
+                var n = notification as IHasFollowUpNotifications;
```
Comment (peppy) [inline osu.Game/Overlays/NotificationOverlay.cs]: what if the "follow up notification" isn't yet decided at the point of posting? what if it's added along the way depending on what happened during the span of the first one?

File: osu.Game/Overlays/Notifications/DrawableNotification.cs [hunk 16]
```diff
@@ -16,15 +16,15 @@
...
+    public abstract class DrawableNotification : Container
     {
         /// <summary>
         /// User requested close.
         /// </summary>
         public event Action Closed;
 
         /// <summary>
-        /// Run on user activating the notification. Return true to close.
+        /// Run on user activating the drawableNotification. Return true to close.
```
Comment (peppy) [inline osu.Game/Overlays/Notifications/DrawableNotification.cs]: ?

**Review by peppy — CHANGES_REQUESTED** [review]
not a full review, stopping short because too many issues

File: osu.Game/Notifications/NotificationIcon.cs (hunk 9 above)
Comment (smoogipoo) [reply osu.Game/Notifications/NotificationIcon.cs]: That isn't enforced anywhere, considering these properties + NotificationIcon are both public.

Nothing stops me from wanting to alter just the background colour and thinking i can do `NotificationIcon.BackgroundColour = ....`

Comment (peppy) [reply osu.Game/Notifications/NotificationIcon.cs]: In my working through this PR in discord, i told him to make these `readonly`. It does need to be enforced.

File: osu.Game/Overlays/NotificationOverlay.cs (hunk 15 above)
Comment (smoogipoo) [reply osu.Game/Overlays/NotificationOverlay.cs]: It's a god thing that the completion action is a delegate, so this scenario would still work. The only one that wouldn't work is if follow up notifications were added after the notification was completed.

File: osu.Game/Notifications/NotificationIcon.cs (hunk 9 above)
Comment (miterosan) [reply osu.Game/Notifications/NotificationIcon.cs]: Get only properties are readonly, aren't they?

File: osu.Desktop/Overlays/VersionManager.cs (hunk 13 above)
Comment (miterosan) [reply osu.Desktop/Overlays/VersionManager.cs]: I do not know what happened with container there. 

So, should i create factory~ish functions  for creating the notifications?

Comment (peppy) [reply osu.Desktop/Overlays/VersionManager.cs]: i don't have a ready solution, just stating i still prefer the old code in the current state.

**Comment by peppy** [comment]
While I agree this is likely the correct direction to take things, this PR is very outdated and will not cleanly merge (and was left in a state where CI wasn't passing). Feel free to reopen if this gets some further work.

**Pull request closed without merging**